### PR TITLE
VIM-7242: Networking refactor

### PR DESCRIPTION
**Pull request by rpassis**
#### Ticket

[VIM-7242](https://vimean.atlassian.net/browse/VIM-7242)

#### Pull Request Checklist

- [X] Resolved any merge conflicts
- [X] No build errors or warnings are introduced
- [X] New files are written in Swift
- [X] New classes contain license headers
- [X] New classes have Documentation
- [X] New public methods have Documentation

#### Issue Summary

Currently we have multiple targets that depend directly on `AFNetworking` for various reasons. These include `VimeoKit`, `VimeoUpload`, `iOS` and `macOS` (to a lesser extent).

The goal of this PR is make the necessary changes to allow us to prevent these targets from consuming `AFNetworking` directly, and only exposing the minimum functionality required via `VimeoNetworking` as a proxy.

#### Implementation Summary

This PR has been broken down into a few different subtickets that have been previously reviewed by various members of the team. The combined changes in this feature branch include the following:

* Introduction of a new `NetworkReachabilityManager` that mirrors the API and functionality previosuly provided by `AFNetworkingReachabilityManager`. This also includes various unit testing classes.

* Small refactor of `VimeoClient` and `VimeoSessionManager` to move some of the request logic to live inside `VimeoSessionManager`.

* New `SessionManaging` & `AuthenticationListeningDelegate` protocols to abstract `VimeoClient`'s direct dependency on `VimeoSessionManager`

* Addition of various new protocols to help abstract our dependency on `AFNetworking` - `EndpointType`, `Cancellable` and `URLRequestConvertible`.

* Addition of new URL and JSON request serializers (currently unused), both thoroughly tested.

#### Reviewer Tips

To help with context and motivation for some of these changes it might be helpful to review the PRs currently open on the frameworks that currently dependend on AFNetworking:

[VimeoKit](https://github.vimeows.com/MobileApps/VimeoKit/pull/473)
[VimeoUpload](https://github.com/vimeo/VimeoUpload/pull/230)
[Vimeo iOS](https://github.vimeows.com/MobileApps/Vimeo-iOS/pull/2972)

#### How to Test

- Build and run, ensure all tests pass
- Switch to the iOS repo and checkout the branch on the PR above, ensure there are no changes in the current app behavior and that all tests pass. In particular please make sure you can upload videos as usual.

### CODE REVIEW

**Comment by cameo-github** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">SwiftLint not installed, download from https://github.com/realm/SwiftLint</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 174 tests, with 0 failures (0 unexpected) in 1.349 (1.664) seconds</td>
    </tr>
  </tbody>
</table>

See build details on [CircleCI](https://circleci.com/gh/vimeo/VimeoNetworking/1462)

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Sources/Shared/Core/SessionManaging.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/SessionManaging.swift]: tiny nit, extra "to" at the end

File: Sources/Shared/Reachability/NetworkReachabilityManager.swift
Comment (nicolelehrer) [inline Sources/Shared/Reachability/NetworkReachabilityManager.swift:78]: Any need to follow up on this before merge?

File: Sources/Shared/VimeoClient.swift
Comment (nicolelehrer) [inline Sources/Shared/VimeoClient.swift:121]: Where do we kick this off now?

Comment (nicolelehrer) [reply Sources/Shared/VimeoClient.swift:121]: Ah ok, it's on init 👍 - `this reachability manager immediately starts listening
    /// for reachability changes and broadcasts them via system notification to
    /// interested subscribers.`

File: Sources/Shared/Core/JSONEncoding.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/JSONEncoding.swift]: nit extra line

File: Sources/Shared/Core/URLEncoding.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/URLEncoding.swift]: Should this be marked private in lieu of marking members?

File: Sources/Shared/Core/URLRequestConvertible.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/URLRequestConvertible.swift:29]: What are the cases when we want to obfuscate `URLRequest`?

File: Sources/Shared/Core/SessionManaging.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/SessionManaging.swift]: Should this protocol have a more specific doc associated with it?

File: Sources/Shared/Request.swift
Comment (nicolelehrer) [inline Sources/Shared/Request.swift:51]: Any chance that a method like `connect` would need to have a more forgiving retry policy?

File: Sources/Shared/Core/SessionManaging.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/SessionManaging.swift]: To make this more flexible should we provide an input parameter of type bool that specifies if pending tasks should be aborted or not? This would let us switch on the URLSession api available

File: Sources/Shared/VimeoSessionManager.swift
Comment (nicolelehrer) [inline Sources/Shared/VimeoSessionManager.swift:318]: Should we add a TODO here or other explanation?

**Review by nicolelehrer — CHANGES_REQUESTED** [review]
All tests pass 🎉 - smoke testing with client integration now but wanted to release comments sooner

File: Sources/Shared/Core/SessionManaging.swift
Comment (rpassis) [reply Sources/Shared/Core/SessionManaging.swift]: Agreed, and this was done on a [separate branch/PR](https://github.com/vimeo/VimeoNetworking/pull/359/files#diff-f2ecce9caba717801d99d9e8f3fbf565R32) along with other changes

File: Sources/Shared/Core/URLRequestConvertible.swift
Comment (rpassis) [reply Sources/Shared/Core/URLRequestConvertible.swift:29]: The API I have in mind for making networking requests is something like this:

```swift
class SessionManager {
    func request(_ urlRequestConvertible: URLRequestConvertible, then callback: ...) { ... }
}
```

Having this `URLRequestConvertible` protocol and the conformances provided means we can use it in different ways:

```swift
let sessionManager = ...
let url = URL(string: ...)
sessionManager.request(url, then: { ... })
```

or

```swift
let sessionManager = ...
let urlRequest = URLRequest(url ...)
urlRequest.method = ...
sessionManager.request(urlRequest, then: { ... })
```

and also

```swift

enum UserEndpoint: EndpointType, URLRequestConvertible {

    case me
    
    var path: String { ... } 
    var method: HTTPMethod { ... }
    var parameters: Parameters? { ... }
    
    func asURLRequest() throws -> URLRequest {
         // build URLRequest here based on values defined above
    }
}

let sessionManager = ...
sessionManager.request(UserEndpoint.me, then: { ... })
```

File: Sources/Shared/Request.swift
Comment (rpassis) [reply Sources/Shared/Request.swift:51]: I have no idea to be honest! But given that none of the new methods are currently used/supported, should we keep the same default for all and reconsider if/when the need to support other methods arise?

**Comment by rpassis** [comment]
Hey @nicolelehrer thanks again for taking a look at this PR.

I think I addressed all of your comments both here and client side so please take another look when you get a chance and let me know what you think!

File: Sources/Shared/Core/URLRequestConvertible.swift
Comment (nicolelehrer) [reply Sources/Shared/Core/URLRequestConvertible.swift:29]: Nice - love the endpoint case

**Review by nicolelehrer — COMMENTED** [review]
👍 looks good @rpassis! considered approved, holding off on formal approval just so we don't accidentally merge before figuring out release strategy

File: Sources/Shared/Reachability/NetworkReachabilityManager.swift
Comment (nicolelehrer) [reply Sources/Shared/Reachability/NetworkReachabilityManager.swift:78]: Reopening just to be sure we want to leave this as a TODO

File: Sources/Shared/Reachability/NetworkReachabilityStatus.swift
Comment (nicolelehrer) [inline Sources/Shared/Reachability/NetworkReachabilityStatus.swift:66]: We should support watchOS here if we're mentioning it on reachability tests

File: Tests/Shared/Reachability/NetworkingReachabilityManagerTests.swift
Comment (nicolelehrer) [inline Tests/Shared/Reachability/NetworkingReachabilityManagerTests.swift]: Should this be `testThatDefaultManager...`?

File: Sources/Shared/VimeoResponseSerializer.swift
Comment (nicolelehrer) [inline Sources/Shared/VimeoResponseSerializer.swift]: Nit - documentation

File: Sources/Shared/VimeoClient.swift
Comment (nicolelehrer) [inline Sources/Shared/VimeoClient.swift:470]: Good call to investigate due to cases like backing out of a request by leaving a page, upload, etc

File: Tests/Factory/VimeoClientFactory.swift
Comment (nicolelehrer) [inline Tests/Factory/VimeoClientFactory.swift:18]: nit - docs for public API

File: Tests/Factory/VimeoClientFactory.swift
Comment (nicolelehrer) [inline Tests/Factory/VimeoClientFactory.swift:38]: Could we potentially isolate `fake` to just the stub `clientIdentifier` and `clientSecret`, since the rest of the configuration is valid? Maybe break out all constants to default params? For example, we would perhaps want it to be apparent what the default API version is.

File: Tests/Factory/VimeoClientFactory.swift
Comment (nicolelehrer) [inline Tests/Factory/VimeoClientFactory.swift]: Should we consider a compiler directive? like #warning("Add your client secret") or too annoying. Perhaps bundling these into a client struct or tuple?

File: Sources/Shared/Core/SessionManaging.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/SessionManaging.swift:12]: Should we consider enforcing passing the request? No strong preference on my end, just a thought

File: Sources/Shared/VimeoSessionManager.swift
Comment (nicolelehrer) [inline Sources/Shared/VimeoSessionManager.swift:100]: nit - Docs for public API?

File: Sources/Shared/VimeoSessionManager.swift
Comment (nicolelehrer) [inline Sources/Shared/VimeoSessionManager.swift:205]: 🔥 🔥 🔥

File: Sources/Shared/VimeoSessionManager.swift
Comment (nicolelehrer) [inline Sources/Shared/VimeoSessionManager.swift:241]: Could we make it explicit that this is for background capable uploads in the documentation?

File: Sources/Shared/Core/VNError.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/VNError.swift]: Do we have to keep the file name short due to linting?

File: Sources/Shared/Core/AuthenticatingListeningDelegate.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/AuthenticatingListeningDelegate.swift]: `AuthenticationListening` or `AuthenticationListener`? or does delegate help infer something else

File: Sources/Shared/Core/Request+EndpointType.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/Request+EndpointType.swift]: nit - explicit self

File: Sources/Shared/Core/Request+EndpointType.swift
Comment (nicolelehrer) [inline Sources/Shared/Core/Request+EndpointType.swift:20]: Note to check in on this part

**Review by nicolelehrer — CHANGES_REQUESTED** [review]
Looking good! Tests pass 👍 Minor things to address

File: Sources/Shared/Reachability/NetworkReachabilityStatus.swift
Comment (rpassis) [reply Sources/Shared/Reachability/NetworkReachabilityStatus.swift:66]: Removed from tests - this is not supported on watchOS

File: Tests/Factory/VimeoClientFactory.swift
Comment (rpassis) [reply Tests/Factory/VimeoClientFactory.swift]: This is only used in the test target and we don't require the correct ID and secret to be set. I will just rename it FAKE_ID / FAKE_SECRET to make it clear that this ok.

File: Sources/Shared/Core/SessionManaging.swift
Comment (rpassis) [reply Sources/Shared/Core/SessionManaging.swift:12]: That would be preferrable for sure but unfortunately at this point we cannot guarantee that a request will exist - we may pass in a `URLRequestConvertible` type which attempts to create a `URLRequest` and throws an error if it fails.

File: Sources/Shared/VimeoSessionManager.swift
Comment (rpassis) [reply Sources/Shared/VimeoSessionManager.swift:241]: As far as I am aware there is nothing in this method that restricts it (or makes it specifically suitable) for background uploads. Isn't this the domain of the upload library?

File: Sources/Shared/Core/VNError.swift
Comment (rpassis) [reply Sources/Shared/Core/VNError.swift]: I renamed the class and forgot to rename the file, fixed!

File: Sources/Shared/Core/AuthenticatingListeningDelegate.swift
Comment (rpassis) [reply Sources/Shared/Core/AuthenticatingListeningDelegate.swift]: Yes good call! At first I wanted to separate SessionManaging and AuthenticationListening as separate types because it felt they are responsible for very different things however after looking at this again I think it makes more sense to simplify this to one type - `SessionManaging` that inherits from `AuthenticationListener`.

File: Sources/Shared/Core/Request+EndpointType.swift
Comment (rpassis) [reply Sources/Shared/Core/Request+EndpointType.swift:20]: I'd like to leave that TODO there for the future as we are unable to fix this with the way things are currently structured.

File: Tests/Factory/VimeoClientFactory.swift
Comment (rpassis) [reply Tests/Factory/VimeoClientFactory.swift:38]: Not sure if I follow this here - also note that this is for the unit test target only so it is definitely a `fake` in this case.

File: Sources/Shared/Reachability/NetworkReachabilityStatus.swift
Comment (rpassis) [reply Sources/Shared/Reachability/NetworkReachabilityStatus.swift:66]: I also added a copy of the `Mutex` type we currently have on `VimeoKit` here to ensure we are safely mutating `previousStatus` on the reachability manager.

**Comment by rpassis** [comment]
Thanks for all the feedback Nicole! I think I addressed almost all of your questions and comments, and left a couple of responses for you to take a look.

File: Sources/Shared/VimeoSessionManager.swift
Comment (nicolelehrer) [reply Sources/Shared/VimeoSessionManager.swift:241]: This method will return an instance of https://developer.apple.com/documentation/foundation/nsurlsessionuploadtask?language=objc which is background capable. So it's possible to support background uploads but it isn't mandated (and depends heavily on the session being configured for background tasks) so on second thought I think you're right in that it's not necessary.

File: Tests/Factory/VimeoClientFactory.swift
Comment (nicolelehrer) [reply Tests/Factory/VimeoClientFactory.swift:38]: Ah now I see this is only for unit tests 👍

**Review by nicolelehrer — APPROVED** [review]
Just need to sync with dev then good to go! 🎉

**Comment by rpassis** [comment]
Thanks @nicolelehrer!

**Pull request merged**
